docs added for taps maintainers

bfontaine opened the pull request:
You can read it at https://github.com/bfontaine/homebrew/blob/doc-tap/share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md

xu-cheng commented:
Maybe worth to mentioned that we prefer that formulae are added into `Formula` directory.

bfontaine commented:
@xu-cheng Thanks, I reworded it as:

> …and can be added under `Formula` or `HomebrewFormula`. It works if there are at the root of the repo but this is not recommended.

dunn commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+elsewhere. If hosted on GitHub, the repo’s name must start with `homebrew-`.
+
+Formulae follow the same format as the core’s ones, and can be added under
+`Formula` or `HomebrewFormula`. It works if there are at the root of the repo
```
"It works if [they] are"

dunn commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+no core formula with the same name, or with `brew install user/repo/foo` to
+avoid conflicts.
+
+Note that if your tap in on GitHub, executing `brew install user/repo/foo` will
```
"if your tap [is] on GitHub"

dunn commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+avoid conflicts.
+
+Note that if your tap in on GitHub, executing `brew install user/repo/foo` will
+automatically install your `github.com/user/homebrew-repo` tap to install the
```
This might be a little clearer as "automatically add your `github.com/user/homebrew-repo` tap before installing the formula"

bfontaine commented:
Thank you @dunn, I fixed everything.

dunn commented:
Should we link to a few popular examples?

tdsmith commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+
+Formulae follow the same format as the core’s ones, and can be added under
+`Formula` or `HomebrewFormula`. It works if they are at the root of the repo
+but this is not recommended.
```
When we offer recommendations like this I think it would be good to explain why. I also think "root of the repo" is fine for most users.

xu-cheng replied:
The reasons are:
* We treat `Formula` folder with highest priority. [Related code](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/formulary.rb#L248-L250)
* It's more organize considering with other folders such as `cmd` and `Casks` which have their own usage.
* It will offer a better view on Github page. For example, it's very hard to see the README for `homebrew/science` since there are so many formulae in the root of the repo.

tdsmith replied:
Most taps will be small and will only contain formulae. It's simpler and (conversely to your third point) makes more visible which formulae a tap provides if the formulae are placed in the root. My preference is to advise new tap owners to put formulae in the repository root, and then say that formulae can also be placed inside a `Formula` or `HomebrewFormula` directory.

bfontaine replied:
This was the original wording. I agree with @xu-cheng regarding the organization even if I often advise people to put formulae at the root of the repo to keep things simple.

tdsmith replied:
I don't think aesthetic differences should cause us to explicitly disrecommend a simpler option but I'll live with whatever you want to do here.

mikemcquaid replied:
I think it's fine to not recommend this too.

bfontaine commented:
@dunn we have some here: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Interesting-Taps-%26-Branches.md

I can include a link in this doc.

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+
+## Creating a tap
+A tap must be a git repository available online. It can be hosted on GitHub or
+elsewhere. If hosted on GitHub, the repo’s name must start with `homebrew-`.
```
This isn't quite the case on GitHub; might just be worth saying `we recommend the repository's name begins with homebrew-`

mikemcquaid replied:
Also use `repository` instead of `repo`

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+elsewhere. If hosted on GitHub, the repo’s name must start with `homebrew-`.
+
+Formulae follow the same format as the core’s ones, and can be added under
+`Formula` or `HomebrewFormula`. It works if they are at the root of the repo
```
`added under ... directories.`

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,42 @@
+where `user` and `repo` will be used to refer to your tap and `<url>` is your
+Git clone URL.
+
+Users can then install your formulae either with `brew install foo` if there’s
```
I'd rather point people towards `brew install user/repo/foo` and not tell them about `brew tap` until afterwards; nicer to indicate they can install from taps with a single command.

bfontaine commented:
@tdsmith, @xu-cheng: I rephrased the “root vs `Formula/`” part:

> Formulae follow the same format as the core’s ones, and can be added at the repository’s root, or under `Formula` or `HomebrewFormula` subdirectories. We recommend the latter options because it makes the repository organisation easier to grasp, and top-level files are not mixed with formulae.

@mikemcquaid: I moved the one-step install at the top, and fixed your other comments.

bfontaine commented:
I added `homebrew/tex` and `homebrew/aliases` as examples.

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,51 @@
+elsewhere. If hosted on GitHub, we recommend that the repository’s name start
+with `homebrew-`.
+
+Formulae follow the same format as the core’s ones, and can be added at the
```
`Tap formulae follow`

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,51 @@
+to provide their own formulae to any Homebrew user.
+
+## Creating a tap
+A tap must be a git repository available online. It can be hosted on GitHub or
```
Technically it doesn't need to be available online or a Git repository.

bfontaine replied:
Fixed:

> A tap is usually a git repository available online, but you can use anything as long as it’s a protocol that git understands.

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,51 @@
+### Installing
+If it’s on GitHub, users can install any of your formulae with
+`brew install user/repo/formula`. Homebrew will automatically add your
+`github.com/user/homebrew-repo` tap before installing the formula.
```
Might be worth clarifying that `user/repo/formula` == `https://github.com/user/homebrew-repo/*/formula.rb`

bfontaine replied:
I added it below:

> `user/repo/formula` points to the `github.com/user/homebrew-repo/**/formula.rb` file here.

DomT4 commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,51 @@
+# How to Create and Maintain a Tap
+Taps are external sources of Homebrew formulae. They can be created by anyone
```
Possibly the world's smallest nit here - You can use taps for commands as well as/instead of formulae.

mikemcquaid replied:
:+1:

bfontaine replied:
@mikemcquaid This is mentioned at the end: https://github.com/Homebrew/homebrew/pull/41858#discussion_r35016491

mikemcquaid replied:
Would be good to integrate here too or have a wee table of contents.

bfontaine replied:
@mikemcquaid: rephrased:

> Taps are external sources of Homebrew formulae and/or external commands. They can be created by anyone to provide their own formulae and/or external commands to any Homebrew user.

DomT4 commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,51 @@
+
+## External commands
+You can provide your tap users with custom `brew` commands by adding them in a
+`cmd` subdirectory. [Read more on external commands](External-Commands.md).
```
:+1:, Feel free to ignore my comment further up :wink:

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,54 @@
+
+## Creating a tap
+A tap is usually a git repository available online, but you can use anything as
+long as it’s a protocol that git understands.
```
`or even just a directory with files in it`

bfontaine replied:
Thanks @mikemcquaid, fixed.

mikemcquaid commented on share/doc/homebrew/How-to-Create-and-Maintain-a-Tap.md:
```diff
@@ -0,0 +1,56 @@
+# How to Create and Maintain a Tap
+Taps are external sources of Homebrew formulae and/or external commands. They
+can be created by anyone to provide their own formulae and/or external commands
+to any Homebrew user.
```
:+1:

bfontaine replied:
@mikemcquaid Do I have your green light to merge this PR? :smile:

mikemcquaid replied:
:green_heart:

mikemcquaid replied:
(:+1:)
